byzcoin/tx_pipeline.go

```diff
@@ -2,263 +2,230 @@ package byzcoin
+		},
+	}
+}
+
+// start listens for new ClientTransactions and queues them up.
+// It also listens for update requests and puts them in the queue.
+// If a block is pending for approval,
+// the next block is already in the channel.
+// New ClientTransactions coming in during this time will still be appended
+// to the block-proposition in the channel, until the proposition is too big.
+func (p *txPipeline) start(currentState *proposedTransactions,
+	stopSignal chan struct{}) {
+
+	// Stores the known transaction-hashes to avoid double-inclusion of the
+	// same transaction.
+	var txHashes [][]byte
+
+	// newBlock also serves as cache for the latest proposedTransactions: if the
+	// new block hasn't been produced, it is legit to read the channel,
+	// update the state, and write it back in.
+	newBlock := make(chan *proposedTransactions, 1)
+	blockSent := make(chan struct{}, 1)
+	p.wg.Add(1)
+	go p.createBlocks(newBlock, blockSent)
+
+leaderLoop:
+	for {
+		select {
+		case <-stopSignal:
+			// Either a view-change or the node goes down.
+			close(newBlock)
+			break leaderLoop
+
+		case <-blockSent:
+			// A block has been proposed and either accepted or rejected.
+			p.newVersion = 0
+
+		case version := <-p.needUpgrade:
+			// An upgrade of the system-version is needed.
+			currVers, err := p.processor.GetVersion()
+			if err != nil {
+				log.Errorf("needUpgrade error: %v", err)
+				continue
+			}
+			if version > currVers {
+				p.newVersion = version
+			}
+
+		case tx := <-p.ctxChan:
+			// A new ClientTransaction comes in - check if it's unique and
+			// put it in the queue if it is.
+			txh := tx.Instructions.HashWithSignatures()
+			for _, txHash := range txHashes {
+				if bytes.Equal(txHash, txh) {
+					log.Lvl2("Got a duplicate transaction, ignoring it")
+					continue leaderLoop
+				}
+			}
+			txHashes = append(txHashes, txh)
+			if len(txHashes) > maxTxHashes {
+				txHashes = txHashes[len(txHashes)-maxTxHashes:]
+			}
+
+			p.txQueue = append(p.txQueue, tx)
+		}
+
+		// Check if a block is pending, fetch it if it's the case
+		select {
+		case nbState, ok := <-newBlock:
+			if ok {
+				currentState = nbState
+			}
+		default:
+		}
+
+		// For a version update, need to recover the ClientTransactions,
+		// and send a version update
+		if p.newVersion > 0 {
+			newBlock <- &proposedTransactions{newVersion: p.newVersion,
+				sst: currentState.sst}
+			if (len(currentState.txs)) > 0 {
+				txs := make([]ClientTransaction, len(currentState.txs))
+				for i, txRes := range currentState.txs {
+					txs[i] = txRes.ClientTransaction
+				}
+				p.txQueue = append(txs, p.txQueue...)
+			}
+			continue
+		}
+
+		// Add as many ClientTransactions as possible to the proposedTransactions
+		// before the block gets too big, then put it in the channel.
+		p.txQueue = currentState.addTransactions(p.processor, p.txQueue)
+		if !currentState.isEmpty() {
+			newBlock <- currentState.copy()
+			currentState.reset()
+		}
+	}
+	p.wg.Wait()
```

Comment:
so, this function never ends, right? why using the `p.wg` and not a IMO more readable `<-make(chan struct{})`?

Reply:
if the `stopSignal` channel is closed, the function quits. That's the case when the leader changes. Even though, in most cases, when the leader changes it's because the leader is down...

So this is probably mostly used when the leader exits, another leader takes over, the old leader starts again, starts `txPipeline.start`, catches up the blocks, sees that he's no longer the leader, and closes `stopSignal`.